src/hotspot/share/nmt/memLogRecorder.hpp

```diff
@@ -0,0 +1,212 @@
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_MEMLOGRECORDER_HPP
+#define SHARE_NMT_MEMLOGRECORDER_HPP
+
+#include "memory/allocation.hpp"
+#include "nmt/nmtCommon.hpp"
+#include "runtime/globals.hpp"
+#include "runtime/java.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "utilities/macros.hpp"
+
+// see MemTracker::NmtVirtualMemoryLocker
+class NMTRecorder_Locker: StackObj {
+  static bool _safe_to_use;
+  ConditionalMutexLocker _cml;
+public:
+  NMTRecorder_Locker(): _cml(NMTRecorder_lock, _safe_to_use, Mutex::_no_safepoint_check_flag) {
+  }
+  ~NMTRecorder_Locker() {
+  }
+  static inline bool is_safe_to_use() {
+    return _safe_to_use;
+  }
+  static inline void set_safe_to_use() {
+    _safe_to_use = true;
+  }
+};
+
+#if defined(LINUX) || defined(__APPLE__)
+
+#if defined(LINUX)
+#define MAXTHREADNAMESIZE 256
+#endif
+
+class NMT_LogRecorder : public StackObj {
+protected:
+  long int _limit  = 0;
+  long int _count  = 0;
+  int _log_fd;
+  volatile bool _done = true;
+
+protected:
+  volatile size_t _threads_names_size = 0;
+  typedef struct thread_name_info {
+    char name[MAXTHREADNAMESIZE];
+    long int thread;
+  } thread_name_info;
+  thread_name_info *_threads_names = nullptr;
+
+public:
+  static void initialize();
+  static void finish();
+  static void replay();
+  static void logThreadName(const char* name);
+
+public:
+  void init();
+  void get_thread_name(char* buf);
+  bool done() {
+    return _done;
+  }
+  void logThreadName();
+};
+
+class NMT_MemoryLogRecorder : public NMT_LogRecorder {
+
+private:
+  static NMT_MemoryLogRecorder _recorder;
+
+private:
+    struct Entry {
+    long int time;
+    long int thread;
+    address ptr;
+    address old;
+    address stack[NMT_TrackingStackDepth];
+    long int requested;
+    long int actual;
+    long int mem_tag;
+  };
+
+public:
+  static NMT_MemoryLogRecorder* instance() {
+    return &_recorder;
+  };
+  static void initialize(long int count);
+  static bool initialized() {
+    return false;
+  }
+  static void print(Entry *e);
+  static void finish(void);
+  static void replay(const int pid);
+  static void record_free(void *ptr);
+  static void record_malloc(MemTag mem_tag, size_t requested, void* ptr, const NativeCallStack *stack, void* old = nullptr);
+  static void printActualSizesFor(const char* list);
+
+private:
+  static void _record(MemTag mem_tag, size_t requested, address ptr, address old, const NativeCallStack *stack);
+};
+
+class NMT_VirtualMemoryLogRecorder : public NMT_LogRecorder {
+
+private:
+  static NMT_VirtualMemoryLogRecorder _recorder;
+
+private:
+  struct Entry {
+    long int time;
+    long int thread;
+    address ptr;
+    address stack[NMT_TrackingStackDepth];
+    long int mem_tag;
+    long int mem_tag_split;
+    size_t size;
+    size_t size_split;
+    int type;
+  };
+
+public:
+  enum Type {
+    RESERVE,
+    RELEASE,
+    UNCOMMIT,
+    RESERVE_AND_COMMIT,
+    COMMIT,
+    SPLIT_RESERVED,
+    TAG
+  };
```

Comment:
Better name than `Type`, like `MemoryOperation`? No need for `ALL_CAPS` names if you don't want to, you can use `ThisTypeOfName` instead. That's a style choice you get to make, though.

Reply:
I like your suggestion, done.